From a6064c936fabbb7a672a07dd17f6800a8e9eac99 Mon Sep 17 00:00:00 2001
From: Mark Syms <mark.syms@citrix.com>
Date: Tue, 26 May 2026 09:15:00 +0100
Subject: [PATCH] CP-312848: allow tapdisk-pause.Pause to be idempotent when
 safe

If we do not have the paused key set in sm_config but the tapdisk is
paused the likely cause is an earlier failed refresh. Allow the new
pause to be issued idempotently in this case to allow for recovery.
diff --git a/drivers/tapdisk-pause b/drivers/tapdisk-pause
index 6b7fc09..9b2e23f 100755
--- a/drivers/tapdisk-pause
+++ b/drivers/tapdisk-pause
@@ -28,10 +28,12 @@ import XenAPI
 import lvhdutil
 import vhdutil
 import lvmcache
+import VDI
 
 TAPDEV_BACKPATH_PFX = "/dev/sm/backend"
 TAPDEV_PHYPATH_PFX = "/dev/sm/phy"
 
+
 def locking(excType, override=True):
     def locking2(op):
         def wrapper(self, *args):
@@ -130,7 +132,12 @@ class Tapdisk:
                     self.realpath = p
                     if ty == "LV": self.vdi_type = "aio"
                     else: self.vdi_type = "vhd"
-        
+
+    def _is_snap_paused(self):
+        vdi_ref = self.session.xenapi.VDI.get_by_uuid(self.vdi_uuid)
+        sm_config = self.session.xenapi.VDI.get_sm_config(vdi_ref)
+        return 'paused' in sm_config
+
     @locking("VDIUnavailable")
     def Pause(self):
         util.SMlog("Pause for %s" % self.vdi_uuid)
@@ -144,6 +151,10 @@ class Tapdisk:
         tapargs = {"minor":self.minor}
         util.SMlog("Calling tap pause with minor %d" % self.minor)
         tapdisk = blktap2.Tapdisk.get(**tapargs)
+        if tapdisk.is_paused() and not self._is_snap_paused():
+            # If we're already paused and not in a snapshot, return
+            return str(True)
+        
         tapdisk.pause()
         return str(True)
 
@@ -171,7 +182,6 @@ class Tapdisk:
                     continue
                 lvm_cache.activate(ns, uuid, lv_name, False)
 
-        import VDI
         vdi = VDI.VDI.from_uuid(self.session, self.vdi_uuid)
         # Check if CBT is enabled on disk we are about to unpause
         if vdi._get_blocktracking_status():
diff --git a/drivers/tapdisk_pause.py b/drivers/tapdisk_pause.py
new file mode 120000
index 0000000..e859b10
--- /dev/null
+++ b/drivers/tapdisk_pause.py
@@ -0,0 +1 @@
+tapdisk-pause
\ No newline at end of file
diff --git a/mocks/XenAPIPlugin/__init__.py b/mocks/XenAPIPlugin/__init__.py
new file mode 100644
index 0000000..e69de29
diff --git a/tests/run_python_unittests.sh b/tests/run_python_unittests.sh
index 6f7317d..e2eb2cc 100755
--- a/tests/run_python_unittests.sh
+++ b/tests/run_python_unittests.sh
@@ -20,9 +20,9 @@ fi
 
 (
     cd "$SMROOT"
-    PYTHONPATH="$SMROOT/mocks:$SMROOT/drivers:$SMROOT/misc/fairlock" \
+    PYTHONPATH="$SMROOT/mocks:$SMROOT/drivers:$SMROOT/misc/fairlock:$SMROOT/scripts/plugins" \
         $COVERAGE run --branch \
-            --source="$SMROOT/drivers,$SMROOT/tests,$SMROOT/misc/fairlock" \
+            --source="$SMROOT/drivers,$SMROOT/tests,$SMROOT/misc/fairlock,$SMROOT/scripts/plugins" \
             -m unittest discover -f -s "$TESTS" -p "$FILES" -v
 
     echo "Test coverage"
diff --git a/tests/test_tapdisk_pause.py b/tests/test_tapdisk_pause.py
new file mode 100644
index 0000000..a003657
--- /dev/null
+++ b/tests/test_tapdisk_pause.py
@@ -0,0 +1,81 @@
+import unittest
+import uuid
+from unittest import mock
+
+import blktap2
+import tapdisk_pause
+
+
+class TestTapdiskPause(unittest.TestCase):
+
+    def setUp(self):
+        self.mock_session = mock.MagicMock()
+        log_patcher = mock.patch("tapdisk_pause.util.SMlog")
+        self.mock_log = log_patcher.start()
+        exists_patcher = mock.patch('tapdisk_pause.os.path.exists')
+        self.mock_exists = exists_patcher.start()
+        lock_patcher = mock.patch('tapdisk_pause.Lock')
+        self.mock_lock = lock_patcher.start()
+        self.mock_tapdisk = mock.create_autospec(blktap2.Tapdisk)
+        blktap2_patcher = mock.patch('tapdisk_pause.blktap2')
+        self.mock_blktap2 = blktap2_patcher.start()
+        self.mock_blktap2.Tapdisk.major.return_value = 254
+        vdi_patcher = mock.patch('tapdisk_pause.VDI')
+        self.mock_vdi = vdi_patcher.start()
+        self.mock_blktap2.Tapdisk.get.return_value = self.mock_tapdisk
+        readlink_patcher = mock.patch('tapdisk_pause.os.readlink')
+        self.mock_readlink = readlink_patcher.start()
+        self.mock_readlink.side_effect = lambda x: x
+        get_dev_patcher = mock.patch('tapdisk_pause._getDevMajor_minor')
+        self.mock_getdev = get_dev_patcher.start()
+        self.addCleanup(mock.patch.stopall)
+
+    def test_refresh_success(self):
+        # Arrange
+        vdi_uuid = str(uuid.uuid4())
+        sr_uuid = str(uuid.uuid4())
+        args = {
+            'vdi_uuid': vdi_uuid,
+            'sr_uuid': sr_uuid
+        }
+
+        self.mock_getdev.return_value = (254, 5)
+        test_vdi = self.mock_vdi.VDI.from_uuid.return_value
+        test_vdi._get_blocktracking_status.return_value = False
+        self.mock_tapdisk.is_paused.return_value = False
+
+        # Act
+        tapdisk_pause.tapRefresh(self.mock_session, args)
+
+        # Assert
+        self.mock_tapdisk.pause.assert_called_once_with()
+        self.mock_tapdisk.unpause.assert_called_once_with(
+            None, None, None, None)
+
+    def test_refresh_recover_from_failed(self):
+        """
+        If a previous refresh has failed after pausing the tapdisk, recover
+        """
+        # Arrange
+        vdi_uuid = str(uuid.uuid4())
+        sr_uuid = str(uuid.uuid4())
+        args = {
+            'vdi_uuid': vdi_uuid,
+            'sr_uuid': sr_uuid
+        }
+
+        self.mock_getdev.return_value = (254, 5)
+        test_vdi = self.mock_vdi.VDI.from_uuid.return_value
+        test_vdi._get_blocktracking_status.return_value = False
+        self.mock_tapdisk.is_paused.return_value = True
+        test_sm_config = {}
+        xapi_session = self.mock_session.xenapi
+        xapi_session.VDI.get_sm_config.return_value = test_sm_config
+
+        # Act
+        tapdisk_pause.tapRefresh(self.mock_session, args)
+
+        # Assert
+        self.mock_tapdisk.pause.assert_not_called()
+        self.mock_tapdisk.unpause.assert_called_once_with(
+            None, None, None, None)
